src/lib.rs

```diff
@@ -0,0 +1,626 @@
+            })?,
+            value,
+        ));
+        let query = self
+            .query_map
+            .lock()
+            .unwrap()
+            .get(&reqid)
+            .ok_or(UStatus::fail_with_code(
+                UCode::Internal,
+                "query doesn't exist",
+            ))?
+            .clone();
+
+        // Send data
+        query
+            .reply(reply)
+            .with_attachment(attachment.build())
+            .map_err(|_| UStatus::fail_with_code(UCode::Internal, "Unable to add attachment"))?
+            .res()
+            .await
+            .map_err(|_| UStatus::fail_with_code(UCode::Internal, "Unable to reply with Zenoh"))?;
+
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl RpcClient for UPClientZenoh {
+    async fn invoke_method(
+        &self,
+        topic: UUri,
+        payload: UPayload,
+        attributes: UAttributes,
+    ) -> RpcClientResult {
+        // Validate UUri
+        UriValidator::validate(&topic)
+            .map_err(|_| RpcMapperError::UnexpectedError(String::from("Wrong UUri")))?;
+
+        // Validate UAttributes
+        {
+            // TODO: Check why the validator doesn't have Send
+            let validator = Validators::Request.validator();
+            if let Err(e) = validator.validate(&attributes) {
+                return Err(RpcMapperError::UnexpectedError(format!(
+                    "Wrong UAttributes {e:?}",
+                )));
+            }
+        }
+
+        // Get Zenoh key
+        let Ok(zenoh_key) = UPClientZenoh::to_zenoh_key_string(&topic) else {
+            return Err(RpcMapperError::UnexpectedError(String::from(
+                "Unable to transform to Zenoh key",
+            )));
+        };
+
+        // Get the data from UPayload
+        let Some(Data::Value(buf)) = payload.data else {
+            // TODO: Assume we only have Value here, no reference for shared memory
+            return Err(RpcMapperError::InvalidPayload(String::from(
+                "Wrong UPayload",
+            )));
+        };
+
+        // Serialized UAttributes into protobuf
+        let mut attr = vec![];
+        let Ok(()) = attributes.encode(&mut attr) else {
+            return Err(RpcMapperError::ProtobufError(String::from(
+                "Unable to encode UAttributes",
+            )));
+        };
+
+        // Add attachment and payload
+        let mut attachment = AttachmentBuilder::new();
+        attachment.insert("uattributes", attr.as_slice());
+        let value = Value::new(buf.into()).encoding(Encoding::WithSuffix(
+            KnownEncoding::AppCustom,
+            payload.format.to_string().into(),
+        ));
+        // TODO: Query should support .encoding
+        // TODO: Adjust the timeout
+        let getbuilder = self
+            .session
+            .get(&zenoh_key)
+            .with_value(value)
+            .with_attachment(attachment.build())
+            .target(QueryTarget::BestMatching)
```

Comment:
@evshary -- seems this is part of what Luca was referring to in the meeting today.

Let's work with a specific scenario
* `Device1` and `Device2` both speak Zenoh as their "host transport"
* `Device1` and `Device2` each have a uStreamer with an instance of `up-client-zenoh-rust` running with `register_listener(specific_remote_UAuthority_either_id_or_ip, callback)`
  * `Device1`: within uStreamer: `up_client_zenoh_rust.register_listener(UAuthority_pointed_to_Device1, zenoh_remote_callback)`
  * `Device1`: within uApp: `up_client_zenoh_rust.register_rpc_listener(specific_uuri, callback)`

Situation:
From `Device2` there's an `invoke_method()` which contains a UAuthority which points to `Device1`

Question:
1. Given that both uApp and uStreamer on `Device1` would under-the-hood register on the query, but uApp's Zenoh KeyExpr is _more specific_ and the uStreamer one  _includes wildcards_ (e.g. `/up/<Device2-UAuthority>/**`) which `QueryTarget` would be chosen -- the uApp one or the uStreamer one?

To be clear -- the desire is the _more specific_ one, i.e. the uApp to be chosen if it's found.

As a follow-up question -- 

2. In the case the Queryable on `Device1` cannot be found, e.g. the case @stevenhartley showed where there is a gateway in between and no direct communication is possible between `Device1` and `Device2`, but on the gateway there is a `up_client_zenoh_rust.register_listener(UAuthority_pointed_to_Device2, zenoh_remote_callback)`, would the `zenoh_remote_callback` from the gateway be seen and triggered?

Reply:
@PLeVasseur 
Let me use Zenoh's way to answer your questions first.

About your first question:
The query target can be set in 3 different modes (BEST_MATCHING, ALL_COMPLETE, ALL)
https://zenoh-c.readthedocs.io/en/latest/api.html#c.z_query_target_t
Assume we have two queryables with `complete` is true (register to `up/uauth/**` and `up/uauth/app` respectively)
Now we want to send query with `up/uauth/app`
* query with `BEST_MATCHING`: Only the nearest one will reply. Might be `up/uauth/**` or `up/uauth/app`
* query with `ALL_COMPLETE` or `ALL`: Both `up/uauth/**` and `up/uauth/app` will reply

If I understand correctly, you want only `up/uauth/app` to reply the query. If there is no `up/uauth/app`, then `up/uauth/**` will take place. Am I right?
If this is your case, maybe @Mallets can give us some suggestions. I'm not pretty sure whether there is a good way to achieve it.

For the second question, the answer is yes.
In fact, if Device1, Device2 and Gateway are all using Zenoh, Device2 can send query to the queryable in Device1 directly. Therefore, maybe in Zenoh case, Gateway can be just a dummy node running in Zenoh peer mode.

Actually, I also had some questions yesterday.
1. If there are two devices both speak Zenoh and can connect to each other directly, do we still need uStreamer to bridge them?
2. If there are two devices (One speak protocolFoo and one speak protocolBar), should both devices need uStreamers?
```
uApp(protocolFoo) ---> uStreamer ---> uStreamer ---> uApp(protocolBar)
 |----------Device 1-----------|      |----------Device 2-----------|
```
Or maybe we just need one uStreamer
```
uApp(protocolFoo) ---> uStreamer ---> uApp(protocolBar)
 |----------Device 1-----------|      |----Device 2----|
```
3. Maybe the most important one. Let me take your scenario as an example.
In your case, while Device2 calls `invoke_method`, uStreamer needs `register_listener` and forward the request to Device1.
However, in the current implementation, `register_listener` can't receive the message from `invoke_method`. We use query & queryable to implement `invoke_method` and `register_rpc_listener`. Besides, we use put & subscribe to implement `send` and `register_listener`. They are different mechanisms.
Maybe I misunderstand your design, but if not, maybe we need to use put and subscribe to implement `invoke_method` and `register_rpc_listener`.

Reply:
Hey @evshary, I'll try to break our discussion down into parts to answer.

> If I understand correctly, you want only `up/uauth/app` to reply the query. If there is no `up/uauth/app`, then `up/uauth/**` will take place. Am I right? If this is your case, maybe @Mallets can give us some suggestions. I'm not pretty sure whether there is a good way to achieve it.

Yup, seems like a not-unreasonable thing to want to accomplish. Favor the fully-complete-no-wildcard key expression. Would be great to hear more on how that can be accomplished.

> For the second question, the answer is yes.

Gotcha. 

> In fact, if Device1, Device2 and Gateway are all using Zenoh, Device2 can send query to the queryable in Device1 directly. Therefore, maybe in Zenoh case, Gateway can be just a dummy node running in Zenoh peer mode.

I'd have to defer to @stevenhartley on how the gateway would be set up, but I think it'd be running a single other transport. So what you suggest may not be possible in that case.

> Actually, I also had some questions yesterday.
> 
> 1. If there are two devices both speak Zenoh and can connect to each other directly, do we still need uStreamer to bridge them?
> 2. If there are two devices (One speak protocolFoo and one speak protocolBar), should both devices need uStreamers?
> 
> ```
> uApp(protocolFoo) ---> uStreamer ---> uStreamer ---> uApp(protocolBar)
>  |----------Device 1-----------|      |----------Device 2-----------|
> ```
> 
> Or maybe we just need one uStreamer
> 
> ```
> uApp(protocolFoo) ---> uStreamer ---> uApp(protocolBar)
>  |----------Device 1-----------|      |----Device 2----|
> ```

Exactly where I was trying to go yesterday :slightly_smiling_face: I guess since `Device1` and `Device2` _may_ have the same host transport, but they also interact with other devices which only support a third `Baz` transport, they'd have to have uStreamers. (An example of this would be if `Device1` and `Device2` also needed to speak over SOME/IP.)

That means in the case though, of sending from `Device1` speaking Zenoh as host transport to `Device2` speaking Zenoh as host transport _for a CE intended for `Device1`_, the uStreamer in `Device1` would be able to simple drop the queryable. Or at least, that's what I'm hoping is possible. And then let the `uApp` on `Device1` service the Zenoh queryable.

That's where I was going with having the query of invoke_method() favor the _most specific_ key expression in my question up above :point_up_2: 

> 3. Maybe the most important one. Let me take your scenario as an example.
>    In your case, while Device2 calls `invoke_method`, uStreamer needs `register_listener` and forward the request to Device1.
>    However, in the current implementation, `register_listener` can't receive the message from `invoke_method`. We use query & queryable to implement `invoke_method` and `register_rpc_listener`. Besides, we use put & subscribe to implement `send` and `register_listener`. They are different mechanisms.
>    Maybe I misunderstand your design, but if not, maybe we need to use put and subscribe to implement `invoke_method` and `register_rpc_listener`.

Right, so I think this is what Luca was saying yesterday, there's a few changes that'd need to be made I think. Take this as a strawman proposal and feel free to implement in a way that makes sense. I'm certainly not a Zenoh expert.

1. Modify uP-L1's `UTransport::send()` impl to read the `UAttributes.type` to then, under the hood, act appropriately
   * If Publish => as you do today
   * If Response => as you do today
   * If Request => dispatch to a new function which can reach the Zenoh queryable setup by uP-L1 helper's `RpcServer::register_rpc_listener()` _or_ (and this is key) the uP-L1 `UTransport::register_listener()` (more on this in next section)
2. Modify uP-L1's `UTransport::register_listener()` impl to have a "normal mode" and a "uStreamer mode" for lack of a better term. What would that look like?
   * If `UTransport::register_listener(uuri, specific_authority_callback)`'s `uuri` matches some condition, then we'd create a special Zenoh key expression that matches all CEs coming from a specific UAuthority. How would we do that?
   * Let's say for now as a proposal that UUri would look like this:
     * ```rust
       UUri {
           authority: {
               name: <doesnt-matter>
               // either ip or id... depending on which is filled
               ip: <some-specific-ip>,  
               id: <some-specific-id>
           },
           entity: {
             name: "*",
             ..Default::default() // i.e. rest of fields don't matter, set to None
           },
           resource: {
             name: "*",
             ..Default::default() // i.e. rest of fields don't matter, set to None
           }
       }
       ```
   * Then if `register_listener(uuri, specific_authority_callback)` saw that specific kind of UUri, it would both:
     * Set up `specific_authority_callback` to trigger on a Zenoh subscription as per normal, _and in addition_
     * Set up `specific_authority_callback` to trigger on a Zenoh query to `up/<some-specific-ip-or-some-specific-id>/**`
     * (Potentially? This is a bit outside of my knowledge...) Set up `specific_authority_callback` to trigger on the case of receiving a Response type request. I'm not sure how that'd work with the Zenoh queryables and how `invoke_method()` and `register_rpc_listener()` are implemented currently
   * so that the uStreamer could operate with only the uP-L1 layer and push the implementation details on how to accomplish this down into each up-client-foo-lang

Hope that makes more sense, but honestly text is sometimes limiting haha. I'd be happy to jump on a Teams call to discuss further and whiteboard to make sure we're on the same page. Lemme know :slightly_smiling_face:

Reply:
Hi @PLeVasseur 

> That means in the case though, of sending from Device1 speaking Zenoh as host transport to Device2 speaking Zenoh as host transport for a CE intended for Device1, the uStreamer in Device1 would be able to simple drop the queryable. Or at least, that's what I'm hoping is possible. And then let the uApp on Device1 service the Zenoh queryable.

I guess uStreamer doesn't support routing messages to another uStreamer, or it might cause some routing issues. (Maybe I'm wrong)
That means uStreamer can know the protocol of both two sides.
Maybe we can make uStreamer ignore the message if it knows both sides are the same protocol.

> Right, so I think this is what Luca was saying yesterday, there's a few changes that'd need to be made I think. Take this as a strawman proposal and feel free to implement in a way that makes sense. I'm certainly not a Zenoh expert.

Good proposal! I'll go through the details and see whether there is another way to fulfill the requirement.
Glad to know that we're on the same page with the issue, and that's why I wrote down the question to double-check with you.

> Hope that makes more sense, but honestly text is sometimes limiting haha. I'd be happy to jump on a Teams call to discuss further and whiteboard to make sure we're on the same page. Lemme know 🙂

Your text is really clear to me now! We can have a discussion on Teams later if there is still something unclear while working on the implementation. Thanks for your clarification.

Reply:
It should only get to the streamer when we cross the protocol boundary, otherwise it can stay within the current technology and address space.

Reply:
Hi @PLeVasseur 
Sorry for the late reply. Thank you for the nice diagram, and I think Luca has answered your questions on the diagram.
We need to store the query while receiving requests and the RPC callback can respond with the stored query.
Here is [the diagram](https://drive.google.com/file/d/1oFoiqwnLJR3V9sBZqb2tu3YkDwvroPrG/view?usp=sharing) I shared yesterday for your reference.
I'm also happy to have a meeting with you for the details, but I'm off until next Wednesday because of the Lunar New Year.
If you're not in hurry, we can discuss it after then.
![uProtocol with Zenoh Design](https://github.com/eclipse-uprotocol/up-client-zenoh-rust/assets/456210/2967cc5e-c5ef-4341-8e60-2113ce05cd5d)

Reply:
> @PLeVasseur let me attempt to explain a bit better how queries work in Zenoh.
> ...
> Let me know if it makes sense. I'm also happy to jump in an ad-hoc meeting to go through the details.

Great, I think I get it. I updated the diagram [here](https://drive.google.com/file/d/1CPyBC5ZZ4eZwB4NpoEayersPUY3qg0T2/view?usp=sharing) to better capture what is going on then.

Seems like the `up-client-zenoh-rust` would do some book-keeping under the hood to allow us when calling `send()` with a `Response` type, it'd reference the held Zenoh Query we got from the original `register_listener()` and then Reply on it. Am I understanding that correctly?
 * This is the `red` flow on the diagram

On the flip side, when there's an incoming Request, we can use `send()` and it'll dispatch under the hood to `send_request()`. Then the `register_listener()` in the uStreamer would be able to "catch" the reply that's coming back. Is that correct?
 * This is the `green` flow on the diagram

> Hi @PLeVasseur
> ...
> Here is [the diagram](https://drive.google.com/file/d/1oFoiqwnLJR3V9sBZqb2tu3YkDwvroPrG/view?usp=sharing) I shared yesterday for your reference.

Probably I'm just not reading the diagram correctly, but I wanted to ensure that I'm understanding this part:
![image](https://github.com/eclipse-uprotocol/up-client-zenoh-rust/assets/11622119/16da1922-2e96-4ed6-aab1-cc462db68e21)

So for every use case other than the uStreamer one where we specify a special UUri which says "give me all remote topics from UAuthority such and such", an app would use either `register_listener` or `register_rpc_listener` as appropriate.

And to confirm: the way you drew the diagram is just meant to mean that a `Receiver` would do _either_ `register_rpc_listener` _or_ `register_listener` and that in the normal case of using `register_listener` on a specific (non-wildcarded) UUri, they would only get Zenoh Subscriptions, and no Zenoh Queryables. Is that right?

Reply:
> > @PLeVasseur let me attempt to explain a bit better how queries work in Zenoh.
> > ...
> > Let me know if it makes sense. I'm also happy to jump in an ad-hoc meeting to go through the details.
> 
> Great, I think I get it. I updated the diagram [here](https://drive.google.com/file/d/1CPyBC5ZZ4eZwB4NpoEayersPUY3qg0T2/view?usp=sharing) to better capture what is going on then.

It looks good to me.

> 
> Seems like the `up-client-zenoh-rust` would do some book-keeping under the hood to allow us when calling `send()` with a `Response` type, it'd reference the held Zenoh Query we got from the original `register_listener()` and then Reply on it. Am I understanding that correctly?
> 
> * This is the `red` flow on the diagram

Correct. 

> 
> On the flip side, when there's an incoming Request, we can use `send()` and it'll dispatch under the hood to `send_request()`. Then the `register_listener()` in the uStreamer would be able to "catch" the reply that's coming back. Is that correct?
> 
> * This is the `green` flow on the diagram

Correct.

> btw @evshary -- from my point of view, we're reaching the "send messages event horizon" and if there are still some things unclear we should tackle them with an online discussion when you're back to make things go faster. We can ask to have this added to the agenda for that week's uProtocol Zenoh meeting.

I'd like to participate as well.

Reply:
> I'd like to participate as well.

I spoke with @stevenhartley offline about giving this agenda item a little room to breathe on our next up-zenoh meeting next Wednesday. Let's tackle there if possible, failing that we could shoot for something offline :+1:

Reply:
Hi @PLeVasseur

> So for every use case other than the uStreamer one where we specify a special UUri which says "give me all remote topics from UAuthority such and such", an app would use either register_listener or register_rpc_listener as appropriate.

Yes. This is my understanding.

> And to confirm: the way you drew the diagram is just meant to mean that a Receiver would do either register_rpc_listener or register_listener and that in the normal case of using register_listener on a specific (non-wildcarded) UUri, they would only get Zenoh Subscriptions, and no Zenoh Queryables. Is that right?

According to the discussion last time, Steven said that we can distinguish whether this is an RPC request based on the format UUri. I think we can choose to create either Zenoh Subscriptions or Zenoh Queryables in the register_listener. So, it's yes to your question.

> btw @evshary -- from my point of view, we're reaching the "send messages event horizon" and if there are still some things unclear we should tackle them with an online discussion when you're back to make things go faster. We can ask to have this added to the agenda for that week's uProtocol Zenoh meeting.

Sorry for missing your messages here. I was just back yesterday. I'm happy to discuss it in the next uProtocol Zenoh meeting, or we can have another meeting if necessary.

Reply:
Hey @evshary -- I'm a little unclear on if we're on the same page. Let's bring it to the weekly uProtocol Zenoh meeting.

I'm sensing there may be one, but it's a bit hard to tease out without chatting.

Reply:
> I'm sensing that I'm missing something, so best we clear it up through a chat.
Sure. It's also ok if you want to have the chat before the weekly meeting.

> We'd need this to listen to Publish, Request, and Response messages being transmitted over the given UAuthority.
Yes, you're right. What I'm talking about is that we can use some API like `validateRpcMethod` to check whether the URI is for RPC or not. If this is for RPC, then we can use queryable. If not, then keep it in the subscriber way.
https://github.com/eclipse-uprotocol/up-spec/blob/main/basics/uri.adoc#3-validators

Reply:
OK, I think I get what you mean. You want to have both queryable and subscriber while working in the uStreamer.
Then maybe we need to have some exceptions in register_listener. For example, while the UURI is for `*`, it'll have both queryable and subscriber.

Reply:
Yup! I think we're almost there.

It should be if the UUri's UAuthority is set to something concrete (no wildcards), but the other two fields of UEntity and UResource are wildcards (or empty could work too), then we'd have a "special" uStreamer mode where all messages, whether Request, Response, or Publish are routed to the caller of that register_listener().

IIRC @stevenhartley  was saying we're not doing "default gateway" routing, so that would mean we shouldn't support the UAuthority = "*" case. Feel free to jump in if I mischaracterized.

Hope makes sense.

Reply:
Hey @evshary :wave: 

Slight revision regarding how the `UUri` should be formatted to get `uStreamer`-like behavior on the callback to get all of: `Request`, `Response`, and `Publish` type of messages:

```rust
UUri {
    authority: {
        name: <doesnt-matter-not-used>,
        number: <some-specific-ip-or-id>,
    },
    ..Default::default()
}
```

Or in other words, only the `UUri.UAuthority.number` matters.

The `register_client()` call would then have the Zenoh key expression be we are registering the callback on (for all of `Request`, `Response`, and `Publish`):

`up/<some-specific-ip-or-id-in-2-digit-hex-bytes>/*`

WDYT? Does that make sense?

Reply:
Hi @PLeVasseur 
Makes sense to me. I guess the default of entity and resource is empty, which should not be used by the normal UURI. Then we can use it as a special way to listen to both put and query.
Let's take this way first and see if it works or not.